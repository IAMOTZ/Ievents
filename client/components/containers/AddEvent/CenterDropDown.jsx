import React from 'react';
import PropTypes from 'prop-types';

const CenterDropDown = props => (
  <select
    id="centers"
    className="form-control ml-md-3"
    name="centerId"
    onChange={props.handleChange}
    defaultValue={props.defaultValue}
  >
    <option value="" name="centerId">choose a center</option>
    {
      props.centers.map(center => (
        <option
          value={center.id}
          key={center.id}
          name="centerId"
        >{center.name}
        </option>))
    }
  </select>
);

CenterDropDown.defaultProps = {
  defaultValue: 0,
  centers: [],
};

/* eslint-disable react/forbid-prop-types */
CenterDropDown.propTypes = {
  centers: PropTypes.array,
  defaultValue: PropTypes.number,
  handleChange: PropTypes.func.isRequired,
};

export default CenterDropDown;
