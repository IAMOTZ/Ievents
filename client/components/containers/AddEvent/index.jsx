import React from 'react';
import PropTypes from 'prop-types';
import { find } from 'lodash';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { addEvent } from '../../../actions/eventActions';
import { stopAsyncProcess } from '../../../actions/commonActions';
import * as asyncProcess from '../../../actions/asyncProcess';
import { validateAddEventInputs } from '../../../helpers/inputValidators';
import './styles.scss';
import View from './View';


@connect((store) => {
  const { user } = store.authReducer;
  const { centers } = store.fetchCentersReducer;
  return {
    userName: user.name,
    userToken: user.token,
    isAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isSuperAdmin: user.role === 'superAdmin',
    centerToBook: find(centers, { id: store.addEventReducer.centerToBook }),
    addingEventStarted: store.addEventReducer.addingEventStarted,
    addingEventResolved: store.addEventReducer.addingEventResolved,
    addingEventError: store.addEventReducer.addingEventError,
  };
})
class AddEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
      inputErrors: {
        titleError: null,
        descriptionError: null,
        dateError: null,
      },
    };
  }

  componentWillMount() {
    const state = { ...this.state };
    const { centerToBook } = this.props;
    state.centerId = centerToBook ? String(centerToBook.id) : null;
    this.setState(state);
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_EVENT));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Store the date selected in the state of this component.
   * @param {String} dateString The date in its string form.
   */
  handleDateSelection = (dateString) => {
    const date = new Date(dateString);
    const formattedDate = `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()}`;
    const state = { ...this.state };
    state.date = formattedDate;
    this.setState(state);
  }

  /**
   * Clears all the inputErros in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      titleError: null, descriptionError: null, dateError: null
    };
    this.setState(state);
  }

  /**
   * Dispatches the action to add the event.
   */
  add = () => {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_EVENT));
    const { title, description, date } = this.state;
    const centerId = this.props.centerToBook.id;
    const eventDetails = {
      title, description, date, centerId,
    };
    const inputErrors = validateAddEventInputs(eventDetails);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(addEvent(eventDetails, this.props.userToken));
      window.scrollTo(0, 0);
    }
  }

  render() {
    if (this.props.addingEventResolved) {
      return <Redirect to="/events" />;
    } else if (!this.props.centerToBook) {
      return <Redirect to="/centers" />;
    }
    return (
      <View
        add={this.add}
        userName={this.props.userName}
        getInput={this.getInput}
        handleDateSelection={this.handleDateSelection}
        isAdmin={this.props.isAdmin}
        dispatch={this.props.dispatch}
        inputErrors={this.state.inputErrors}
        isSuperAdmin={this.props.isSuperAdmin}
        addingEventStarted={this.props.addingEventStarted}
        addingEventError={this.props.addingEventError}
        centerToBook={this.props.centerToBook}
      />
    );
  }
}

AddEvent.defaultProps = {
  userName: '',
  userToken: '',
  isAdmin: false,
  isSuperAdmin: false,
  centerToBook: null,
  addingEventStarted: false,
  addingEventResolved: false,
  addingEventError: '',
  dispatch: () => { },
};

/* eslint-disable react/forbid-prop-types */
AddEvent.propTypes = {
  userName: PropTypes.string,
  userToken: PropTypes.string,
  isAdmin: PropTypes.bool,
  isSuperAdmin: PropTypes.bool,
  centerToBook: PropTypes.object,
  addingEventStarted: PropTypes.bool,
  addingEventResolved: PropTypes.bool,
  addingEventError: PropTypes.string,
  dispatch: PropTypes.func,
};

export default AddEvent;
