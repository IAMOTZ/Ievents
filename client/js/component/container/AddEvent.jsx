import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
// Actions
import { getAllCenters } from '../../actions/centerActions';
import { addEvent, clearStatus } from '../../actions/eventActions';
import { validateAddEventInputs } from '../../helpers/inputValidators';
// Common Components
import UserSideNav from '../common/SideNavigation.jsx';
import CenterOptions from '../common/CenterDropDown.jsx';
import Header from '../common/Header.jsx';
import { UserTopNav } from '../common/TopNavigation.jsx';
import { LoadingIcon } from '../common/LoadingAnimation.jsx';
import { BigAlert, SmallAlert } from '../common/Alert.jsx';

@connect(store => (
  {
    user: store.user.user,
    authenticated: store.user.status.fetched,
    isAdmin: (store.user.user.role === 'admin'),
    isSuperAdmin: (store.user.user.role === 'superAdmin'),
    centers: store.centers.centers,
    defaultCenter: store.centers.toBook,
    status: {
      error: store.events.status.addingError.message,
      success: store.events.status.added,
      adding: store.events.status.adding,
    },
  }
))
export default class AddEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
      inputErrors: {
        titleError: null,
        descriptionError: null,
        dateError: null,
        centerIdError: null,
      },
    };
  }

  componentWillMount() {
    this.props.dispatch(getAllCenters());
    this.setState({ centerId: this.props.defaultCenter });
  }

  componentWillUnmount() {
    this.props.dispatch(clearStatus('ALL'));
  }

  /**
   * Update some state variables with the user inputs.
   * @param {Event} e The event object.
   */
  getInput = (e) => {
    const state = { ...this.state };
    state[e.target.name] = e.target.value;
    this.setState(state);
  }

  /**
   * Clears all the inputErros in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      titleError: null,
      descriptionError: null,
      dateError: null,
      centerIdError: null,
    };
    this.setState(state);
  }

  /**
   * Dispatches the action to add the event.
   */
  add = () => {
    const {
      title, description, centerId,
    } = this.state;
    const date = this.state.date ? this.state.date.replace(/-/g, '/') : null;
    const eventDetails = {
      title, description, date, centerId,
    };
    const inputErrors = validateAddEventInputs(eventDetails);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(addEvent(eventDetails, this.props.user.token));
      window.scrollTo(0, 0);
    }
  }

  render() {
    let component;
    if (!this.props.authenticated) {
      component = (<Redirect to="/users/login" />);
    } else if (this.props.status.success) {
      component = (<Redirect to="/events" />);
    } else {
      component = (
        <div className="add-event-container">
          {/* Top navigation on small screen */}
          <UserTopNav
            name={this.props.user.name}
            title="Add Event"
            isAdmin={this.props.isAdmin}
            isSuperAdmin={this.props.isSuperAdmin}
            dispatch={this.props.dispatch}
          />
          <div className="container-fluid">
            <div className="row">
              {/*  Side navigation on large screen */}
              <UserSideNav
                name={this.props.user.name}
                isAdmin={this.props.isAdmin}
                isSuperAdmin={this.props.isSuperAdmin}
                dispatch={this.props.dispatch}
              />
              {/* Main content */}
              <div className="col-lg-10 offset-md-2" id="main-content">
                {/* Content Header(navigation) on large screen */}
                <Header text="Add Event" />
                {/* Input form */}
                <form className="mt-lg-5 w-lg-50">
                  <LoadingIcon start={this.props.status.adding} size={3} />
                  <BigAlert message={this.props.status.error} />
                  <div className="form-group">
                    <label htmlFor="title">Title</label>
                    <input
                      type="text"
                      id="title"
                      name="title"
                      className="form-control"
                      placeholder="A short description of your event"
                      onChange={this.getInput}
                    />
                    <small
                      id="emailHelp"
                      className="form-text text-muted"
                    >Between 5 and 30 characters
                    </small>
                    <SmallAlert message={this.state.inputErrors.titleError} />
                  </div>
                  <div className="form-group">
                    <label htmlFor="description">Description</label>
                    <textarea
                      rows="6"
                      name="description"
                      id="description"
                      className="form-control"
                      placeholder="More details about the event"
                      onChange={this.getInput}
                    />
                    <small
                      id="emailHelp"
                      className="form-text text-muted"
                    >Less than 200 characters
                    </small>
                    <SmallAlert message={this.state.inputErrors.descriptionError} />
                  </div>
                </form>
                <form className="my-3 form-inline">
                  <div className="form-group">
                    <label htmlFor="date">Date</label>
                    <input
                      type="date"
                      id="date"
                      className="form-control ml-sm-3"
                      name="date"
                      onChange={this.getInput}
                    />
                    <SmallAlert message={this.state.inputErrors.dateError} />
                  </div>
                  <div className="form-group ml-md-3">
                    <label htmlFor="centers">Choose a Center</label>
                    <select
                      id="centers"
                      className="form-control ml-md-3"
                      name="centerId"
                      onChange={this.getInput}
                      defaultValue={this.props.defaultCenter}
                    >
                      <option
                        value=""
                        name="centerId"
                      >choose a center
                      </option>
                      <CenterOptions centers={this.props.centers} />
                    </select>
                    <SmallAlert message={this.state.inputErrors.centerIdError} />
                  </div>
                </form>
                <button
                  className="btn btn-outline-dark"
                  disabled={this.props.status.adding}
                  onClick={this.add}
                >Add
                </button>
              </div>
            </div>
          </div>
          <footer className="d-block d-sm-none mt-5">
            <div className="container text-white text-center py-5">
              <h1>Ievents</h1>
              <p>Copyright &copy; 2017</p>
            </div>
          </footer>
        </div>
      );
    }
    return component;
  }
}
