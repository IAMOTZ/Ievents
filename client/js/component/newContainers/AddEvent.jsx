import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { getAllCenters } from '../../actions/centerActions';
import { addEvent } from '../../actions/eventActions';
import { stopAsyncProcess } from '../../actions/commonActions';
import * as asyncProcess from '../../actions/asyncProcess';
import { validateAddEventInputs } from '../../helpers/inputValidators';
import UserSideNav from '../common/SideNavigation';
import CenterOptions from '../common/CenterDropDown';
import Header from '../common/Header';
import { UserTopNav } from '../common/TopNavigation';
import { LoadingIcon } from '../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../common/Alert';

@connect((store) => {
  const { user } = store.authReducer;
  return {
    user,
    isAdmin: user.role === 'admin',
    isSuperAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isUserAuthenticaed: store.authReducer.loggingUserResolved,
    centers: store.fetchCentersReducer.centers,
    centerToBook: store.addEventReducer.centerToBook,
    addingEventStarted: store.addEventReducer.addingEventStarted,
    addingEventResolved: store.addEventReducer.addingEventResolved,
    addingEventError: store.addEventReducer.addingEventError,
  };
})
class AddEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
      inputErrors: {
        titleError: null,
        descriptionError: null,
        dateError: null,
        centerIdError: null,
      },
    };
  }

  componentWillMount() {
    this.props.dispatch(getAllCenters());
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_EVENT));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Clears all the inputErros in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      titleError: null,
      descriptionError: null,
      dateError: null,
      centerIdError: null,
    };
    this.setState(state);
  }

  /**
   * Dispatches the action to add the event.
   */
  add = () => {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_EVENT));
    const {
      title, description, centerId,
    } = this.state;
    const date = this.state.date ? this.state.date.replace(/-/g, '/') : null;
    const eventDetails = {
      title, description, date, centerId,
    };
    const inputErrors = validateAddEventInputs(eventDetails);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(addEvent(eventDetails, this.props.user.token));
      window.scrollTo(0, 0);
    }
  }

  render() {
    if (!this.props.isUserAuthenticaed) {
      return (<Redirect to="/users/login" />);
    } else if (this.props.addingEventResolved) {
      return (<Redirect to="/events" />);
    }
    return (
      <div className="add-event-container">
        {/* Top navigation on small screen */}
        <UserTopNav
          name={this.props.user.name}
          title="Add Event"
          isAdmin={this.props.isAdmin}
          isSuperAdmin={this.props.isSuperAdmin}
          dispatch={this.props.dispatch}
        />
        <div className="container-fluid">
          <div className="row">
            {/*  Side navigation on large screen */}
            <UserSideNav
              name={this.props.user.name}
              isAdmin={this.props.isAdmin}
              isSuperAdmin={this.props.isSuperAdmin}
              dispatch={this.props.dispatch}
            />
            {/* Main content */}
            <div className="col-lg-10 offset-md-2" id="main-content">
              {/* Content Header(navigation) on large screen */}
              <Header text="Add Event" />
              {/* Input form */}
              <form className="mt-lg-5 w-lg-50">
                <LoadingIcon start={this.props.addingEventStarted} size={3} />
                <BigAlert message={this.props.addingEventError} />
                <div className="form-group">
                  <label htmlFor="title">Title</label>
                  <input
                    type="text"
                    id="title"
                    name="title"
                    className="form-control"
                    placeholder="A short description of your event"
                    onChange={this.getInput}
                  />
                  <small
                    id="emailHelp"
                    className="form-text text-muted"
                  >Between 5 and 30 characters
                    </small>
                  <SmallAlert message={this.state.inputErrors.titleError} />
                </div>
                <div className="form-group">
                  <label htmlFor="description">Description</label>
                  <textarea
                    rows="6"
                    name="description"
                    id="description"
                    className="form-control"
                    placeholder="More details about the event"
                    onChange={this.getInput}
                  />
                  <small
                    id="emailHelp"
                    className="form-text text-muted"
                  >Less than 200 characters
                    </small>
                  <SmallAlert message={this.state.inputErrors.descriptionError} />
                </div>
              </form>
              <form className="my-3 form-inline">
                <div className="form-group">
                  <label htmlFor="date">Date</label>
                  <input
                    type="date"
                    id="date"
                    className="form-control ml-sm-3"
                    name="date"
                    onChange={this.getInput}
                  />
                  <SmallAlert message={this.state.inputErrors.dateError} />
                </div>
                <div className="form-group ml-md-3">
                  <label htmlFor="centers">Choose a Center</label>
                  <select
                    id="centers"
                    className="form-control ml-md-3"
                    name="centerId"
                    onChange={this.getInput}
                    defaultValue={this.props.centerToBook}
                  >
                    <option
                      value=""
                      name="centerId"
                    >choose a center
                      </option>
                    <CenterOptions centers={this.props.centers} />
                  </select>
                  <SmallAlert message={this.state.inputErrors.centerIdError} />
                </div>
              </form>
              <button
                className="btn btn-outline-dark"
                disabled={this.props.addingEventStarted}
                onClick={this.add}
              >Add
                </button>
            </div>
          </div>
        </div>
        <footer className="d-block d-sm-none mt-5">
          <div className="container text-white text-center py-5">
            <h1>Ievents</h1>
            <p>Copyright &copy; 2017</p>
          </div>
        </footer>
      </div>
    );
  }
}

export default AddEvent;
