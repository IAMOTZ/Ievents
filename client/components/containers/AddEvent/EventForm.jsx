import React from 'react';
import PropTypes from 'prop-types';
import Aux from '../../hoc/Aux';
import CenterDropDown from './CenterDropDown';
import { LoadingIcon } from '../../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../../common/Alert';

const EventForm = (props) => {
  const toUpdate = { ...props.toUpdate };
  return (
    <Aux>
      <form className="mt-lg-5 w-lg-50">
        <LoadingIcon start={props.addingEventStarted || props.updatingEventStarted} size={3} />
        <BigAlert message={props.addingEventError || props.updatingEventError} />
        <div className="form-group">
          <label htmlFor="title">Title</label>
          <input
            type="text"
            id="title"
            name="title"
            className="form-control"
            placeholder="A short description of your event"
            defaultValue={toUpdate.title}
            onChange={props.getInput}
          />
          <small
            id="emailHelp"
            className="form-text text-muted"
          >Between 5 and 30 characters
          </small>
          <SmallAlert message={props.inputErrors.titleError} />
        </div>
        <div className="form-group">
          <label htmlFor="description">Description</label>
          <textarea
            rows="6"
            name="description"
            id="description"
            className="form-control"
            placeholder="More details about the event"
            defaultValue={toUpdate.description}
            onChange={props.getInput}
          />
          <small
            id="emailHelp"
            className="form-text text-muted"
          >Less than 200 characters
          </small>
          <SmallAlert message={props.inputErrors.descriptionError} />
        </div>
      </form>
      <form className="my-3 form-inline">
        <div className="form-group">
          <label htmlFor="date">Date</label>
          <input
            type="date"
            id="date"
            className="form-control ml-sm-3"
            name="date"
            defaultValue={toUpdate.title ? toUpdate.date.replace(/\//g, '-') : null}
            onChange={props.getInput}
          />
          <SmallAlert message={props.inputErrors.dateError} />
        </div>
        <div className="form-group ml-md-3">
          <label htmlFor="centers">Choose a Center</label>
          <CenterDropDown
            centers={props.centers}
            handleChange={props.getInput}
            defaultValue={props.centerToBook || toUpdate.centerId}
          />
          <SmallAlert message={props.inputErrors.centerIdError} />
        </div>
      </form>
      <button
        className="btn btn-outline-dark"
        disabled={props.addingEventStarted || props.updatingEventStarted}
        onClick={props.add || props.update}
      >{toUpdate.title ? 'Update' : 'Add'}
      </button>
    </Aux>
  );
};

EventForm.defaultProps = {
  addingEventStarted: undefined,
  addingEventError: '',
  updatingEventStarted: undefined,
  updatingEventError: '',
  centerToBook: 0,
  inputErrors: {},
  toUpdate: {},
  add: undefined,
  update: undefined,
};

/* eslint-disable react/forbid-prop-types */
EventForm.propTypes = {
  addingEventStarted: PropTypes.bool,
  addingEventError: PropTypes.string,
  updatingEventStarted: PropTypes.bool,
  updatingEventError: PropTypes.string,
  getInput: PropTypes.func.isRequired,
  inputErrors: PropTypes.object,
  centers: PropTypes.array.isRequired,
  centerToBook: PropTypes.number,
  toUpdate: PropTypes.object,
  add: PropTypes.func,
  update: PropTypes.func,
};

export default EventForm;
