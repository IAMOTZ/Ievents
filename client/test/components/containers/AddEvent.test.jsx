import React from 'react';
import AddEvent from '../../../components/containers/AddEvent';
import AddEventView from '../../../components/containers/AddEvent/View';
import EventForm from '../../../components/containers/AddEvent/EventForm';

describe('<AddEvent />', () => {
  const props = {
    userName: 'tester',
    userToken: 'justASampleToken',
    isAdmin: false,
    isSuperAdmin: false,
    centerToBook: { id: 5 },
    addingEventStarted: false,
    addingEventResolved: false,
    addingEventError: null,
    dispatch: jest.fn()
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<AddEvent {...props} />);
    });
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to the events page if adding events resolved', () => {
      wrapper.setProps(alterProps({ addingEventResolved: true }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to the centers page if there is no center to book', () => {
      wrapper.setProps(alterProps({ centerToBook: null }));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Behaviour:', () => {
    let wrapper; let wrapperInstance;
    beforeEach(() => {
      props.dispatch.mockReset();
      wrapper = shallow(<AddEvent {...props} />);
      wrapperInstance = wrapper.instance();
    });
    // The names on the following test block refers
    // to the instance method that handles the behaviour tested.
    describe('COMOPONENT_WILL_MOUNT', () => {
      it('should set the id of the center to book in the state', () => {
        expect(wrapper.state('centerId')).toEqual(String(props.centerToBook.id));
      });
    });

    describe('GET_INPUT', () => {
      it('should update the state with user input', () => {
        const userInput = 'cool event';
        const fakeEvent = { target: { name: 'title', value: userInput } };
        wrapperInstance.getInput(fakeEvent);
        expect(wrapper.state('title')).toEqual(userInput);
      });
    });

    describe('CLEAR_INPUT_ERRORS', () => {
      it('should clear input errors in the state', () => {
        const titleError = 'Event title is required';
        wrapper.setState({ inputErrors: { titleError } });
        wrapperInstance.clearInputErrors();
        expect(wrapper.state('inputErrors').titleError).toBeNull();
      });
    });

    describe('HANDLE_DATE_SELECTION', () => {
      it('should update the state with date input being formatted', () => {
        const date = new Date();
        const dateString = date.toDateString();
        const formattedDate = `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()}`;
        wrapperInstance.handleDateSelection(dateString);
        expect(wrapper.state('date')).toEqual(formattedDate);
      });
    });

    describe('ADD', () => {
      it('should dispatch an action and set inputErros if validation fails', () => {
        const emptyTitle = '';
        wrapper.setState({ title: emptyTitle });
        props.dispatch.mockReset();
        wrapperInstance.add();
        expect(typeof wrapper.state('inputErrors').titleError).toEqual('string');
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
      it('should dispatch two acitons and clear inputErrors if validation pass', () => {
        const correctEventData = {
          title: 'test event',
          description: 'this is the test event',
          date: '2018/06/10',
        };
        const previousError = 'Title is required';
        wrapper.setState({
          ...correctEventData, inputErrors: { titleError: previousError }
        });
        props.dispatch.mockReset();
        wrapperInstance.add();
        expect(wrapper.state('inputErrors').titleError).toBeNull();
        expect(props.dispatch.mock.calls.length).toEqual(2);
      });
    });

    describe('COMOPONENT_WILL_UNMOUNT', () => {
      it('should dispatch an action', () => {
        props.dispatch.mockReset();
        wrapper.unmount();
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });
  });
});

describe('<AddEventView />', () => {
  const props = {
    getInput: () => { },
    inputErrors: {},
    userName: 'tester',
    isAdmin: false,
    isSuperAdmin: false,
    dispatch: () => { },
    add: () => { },
    addingEventStarted: false,
    addingEventError: null,
    update: () => { },
    updating: false,
    updatingEventStarted: false,
    updatingEventError: null,
    centerToBook: {},
    eventToUpdate: {},
    handleDateSelection: () => { },
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<AddEventView {...props} />);
    });
    it('should render correctly when adding', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when updating', () => {
      wrapper.setProps(alterProps({ updating: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('<EventForm />', () => {
  const props = {
    addingEventStarted: false,
    updatingEventStarted: false,
    getInput: () => {},
    inputErrors: {},
    eventToUpdate: {},
    add: () => {},
    update: () => {},
    updating: false,
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<EventForm {...props} />);
    });
    it('should render correctly when adding', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when updating', () => {
      wrapper.setProps(alterProps({ updating: true }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should disable the create/update button when updating/adding event starts', () => {
      wrapper.setProps(alterProps({ addingEventStarted: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});
