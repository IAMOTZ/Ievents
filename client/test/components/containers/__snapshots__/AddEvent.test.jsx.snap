// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddEvent /> Rendering: should redirect to the centers page if there is no center to book 1`] = `
<Redirect
  push={false}
  to="/centers"
/>
`;

exports[`<AddEvent /> Rendering: should redirect to the events page if adding events resolved 1`] = `
<Redirect
  push={false}
  to="/events"
/>
`;

exports[`<AddEvent /> Rendering: should render correctly 1`] = `
<View
  add={[Function]}
  addingEventError={null}
  addingEventStarted={false}
  centerToBook={
    Object {
      "id": 5,
    }
  }
  dispatch={[MockFunction]}
  eventToUpdate={Object {}}
  getInput={[Function]}
  handleDateSelection={[Function]}
  inputErrors={
    Object {
      "dateError": null,
      "descriptionError": null,
      "titleError": null,
    }
  }
  isAdmin={false}
  isSuperAdmin={false}
  update={[Function]}
  updating={false}
  updatingEventError=""
  updatingEventStarted={false}
  userName="tester"
/>
`;

exports[`<AddEventView /> Rendering should render correctly when adding 1`] = `
<div
  className="add-event-container"
>
  <AuthTopNavigation
    dispatch={[Function]}
    isAdmin={false}
    isSuperAdmin={false}
    name="tester"
    title="Create Event"
  />
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <SideNavigation
        dispatch={[Function]}
        isAdmin={false}
        isSuperAdmin={false}
        name="tester"
      />
      <div
        className="col-lg-10 offset-md-2"
      >
        <Header
          text="Create Event"
        />
        <div
          className="page-content"
        >
          <LoadingIcon
            size={3}
            start={false}
          />
          <BigAlert
            message={null}
          />
          <div
            className="text-center"
          >
            <p
              className="event-center-name"
            >
              You are booking 
              <span
                className="text-capitalize"
              />
            </p>
          </div>
          <div
            className="inputs-container"
          >
            <EventForm
              add={[Function]}
              addingEventError={null}
              addingEventStarted={false}
              eventToUpdate={Object {}}
              getInput={[Function]}
              inputErrors={Object {}}
              update={[Function]}
              updating={false}
              updatingEventError={null}
              updatingEventStarted={false}
            />
            <div
              className="calendar-container"
            >
              <label
                className="d-block"
              >
                Select the event date
              </label>
              <SmallAlert />
              <Calendar
                handleDateSelection={[Function]}
                selectedDate={null}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    className="d-block d-sm-none mt-5"
  >
    <Footer />
  </span>
</div>
`;

exports[`<AddEventView /> Rendering should render correctly when updating 1`] = `
<div
  className="add-event-container"
>
  <AuthTopNavigation
    dispatch={[Function]}
    isAdmin={false}
    isSuperAdmin={false}
    name="tester"
    title="Edit Event"
  />
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <SideNavigation
        dispatch={[Function]}
        isAdmin={false}
        isSuperAdmin={false}
        name="tester"
      />
      <div
        className="col-lg-10 offset-md-2"
      >
        <Header
          text="Edit Event"
        />
        <div
          className="page-content"
        >
          <LoadingIcon
            size={3}
            start={false}
          />
          <BigAlert
            message={null}
          />
          <div
            className="text-center"
          >
            <p
              className="event-center-name"
            >
              You are booking 
              <span
                className="text-capitalize"
              />
            </p>
          </div>
          <div
            className="inputs-container"
          >
            <EventForm
              add={[Function]}
              addingEventError={null}
              addingEventStarted={false}
              eventToUpdate={Object {}}
              getInput={[Function]}
              inputErrors={Object {}}
              update={[Function]}
              updating={true}
              updatingEventError={null}
              updatingEventStarted={false}
            />
            <div
              className="calendar-container"
            >
              <label
                className="d-block"
              >
                Select the event date
              </label>
              <SmallAlert />
              <Calendar
                handleDateSelection={[Function]}
                selectedDate={null}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    className="d-block d-sm-none mt-5"
  >
    <Footer />
  </span>
</div>
`;

exports[`<EventForm /> Rendering should disable the create/update button when updating/adding event starts 1`] = `
<div
  className="input-form"
>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="A short description of your event"
        type="text"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Between 5 and 30 characters
      </small>
      <SmallAlert />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        className="form-control"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="More details about the event"
        rows="6"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Less than 200 characters
      </small>
      <SmallAlert />
    </div>
  </form>
  <div
    className="action-btns"
  >
    <button
      className="btn ie-blue-button"
      disabled={true}
      onClick={[Function]}
    >
      Create Event
    </button>
    <Link
      className="custom-blue-text small-text"
      replace={false}
      to="centers"
    >
      Choose another center
    </Link>
  </div>
</div>
`;

exports[`<EventForm /> Rendering should render correctly when adding 1`] = `
<div
  className="input-form"
>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="A short description of your event"
        type="text"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Between 5 and 30 characters
      </small>
      <SmallAlert />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        className="form-control"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="More details about the event"
        rows="6"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Less than 200 characters
      </small>
      <SmallAlert />
    </div>
  </form>
  <div
    className="action-btns"
  >
    <button
      className="btn ie-blue-button"
      disabled={false}
      onClick={[Function]}
    >
      Create Event
    </button>
    <Link
      className="custom-blue-text small-text"
      replace={false}
      to="centers"
    >
      Choose another center
    </Link>
  </div>
</div>
`;

exports[`<EventForm /> Rendering should render correctly when updating 1`] = `
<div
  className="input-form"
>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="A short description of your event"
        type="text"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Between 5 and 30 characters
      </small>
      <SmallAlert />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        className="form-control"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="More details about the event"
        rows="6"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Less than 200 characters
      </small>
      <SmallAlert />
    </div>
  </form>
  <div
    className="action-btns"
  >
    <button
      className="btn ie-blue-button"
      disabled={false}
      onClick={[Function]}
    >
      Update Event
    </button>
  </div>
</div>
`;
